pages/index.tsx

```diff
@@ -1,63 +1,179 @@
+import MuxUploader from '@mux/mux-uploader-react';
+import type { MuxUploaderProps } from '@mux/mux-uploader-react';
+import useSwr from 'swr';
 import Link from 'next/link';
-import { breakpoints } from '../style-vars';
-import Layout from '../components/layout';
 import Button from '../components/button';
-import UploadProgressFullpage from '../components/upload-progress-fullpage';
+import Layout from '../components/layout';
+import { breakpoints } from '../style-vars';
+
+const fetcher = (url: string) => fetch(url).then((res) => res.json());
 
 type Props = null;
 
+type ChunkInfo = {
+  size: number;
+  uploadStarted: number;
+  uploadFinished?: number;
+}
+
+type UploadTelemetry = {
+  fileSize: number;
+  uploadStarted: number;
+  uploadFinished?: number;
+  chunkSize: number;
+  chunks: ChunkInfo[];
+};
+
 const Index: React.FC<Props> = () => {
-  const [file, setFile] = useState<File | null>(null);
-  const [showUploadPage, setShowUploadPage] = useState(true);
-  const inputRef = useRef<HTMLInputElement | null>(null);
-
-  const onDrop = useCallback((acceptedFiles) => {
-    if (acceptedFiles && acceptedFiles[0]) {
-      setFile(acceptedFiles[0]);
-      setShowUploadPage(true);
-    } else {
-      console.warn('got a drop event but no file'); // eslint-disable-line no-console
-    }
-  }, []);
+  const [isUploading, setIsUploading] = useState<boolean>(false);
+  const [uploadId, setUploadId] = useState<string>('');
+  const [isPreparing, setIsPreparing] = useState<boolean>(false);
+  const [errorMessage, setErrorMessage] = useState<string>('');
+  const [uploadAnalytics, setUploadAnalytics] = useState<Partial<UploadTelemetry> & Pick<UploadTelemetry, 'chunks'>>({ chunks: []});
 
-  const onInputChange = () => {
-    if (inputRef.current && inputRef.current.files && inputRef.current.files[0]) {
-      setFile(inputRef.current.files[0]);
-      setShowUploadPage(true);
+  const { data } = useSwr(
+    () => (isPreparing ? `/api/uploads/${uploadId}` : null),
+    fetcher,
+    { refreshInterval: 5000 }
+  );
+
+  const upload = data && data.upload;
+
+  const createUpload = async () => {
+    try {
+      return fetch('/api/uploads', {
+        method: 'POST',
+      })
+        .then((res) => res.json())
+        .then(({ id, url }) => {
+          setUploadId(id);
+          return url;
+        });
+    } catch (e) {
+      console.error('Error in createUpload', e);
+      setErrorMessage('Error creating upload.');
+      return Promise.reject(e);
     }
   };
 
-  if (file && showUploadPage) {
-    return <UploadProgressFullpage file={file} resetPage={() => setShowUploadPage(false)}/>;
+  const handleUpload: MuxUploaderProps['onUploadStart'] = ({ detail }) => {
+    setIsUploading(true);
+
+    const initialUploadAnalytics: UploadTelemetry = {
+      fileSize: detail.file.size,
+      chunkSize: detail.chunkSize,
+      uploadStarted: Date.now(),
+      chunks: [],
+    };
+
+    setUploadAnalytics(initialUploadAnalytics);
+  };
```

Comment:
(non-blocking)

The telemetry code is duplicated both here and in the `UploadProgressFullPage` component. That's definitely not great.

There's some other duplication too like the code for `createUpload`.

I think for the sake of getting this merged and the POC with MuxUploader in the wild, we can live with it. The core issue right now seems to be:

- File upload uses `<mux-uploader>`
- Recording in browser uses `Upchunk` directly

I *think* the right solution here is to update `<mux-uploader>` to handle an arbitrary `File` instance - @cjpillsbury what do you think?

If we did that, then the in-browser recording functionality could use `<mux-uploader>` and we don't have to have this duplicated code.

I would say we should get that detail nailed down by the time we're GA with Mux Uploader.

Let's do that in a follow-up?

Reply:
I agree with everything you said here. Should be a quick lift (🤞) to add direct `file` support. _**Technically**_ we can already do this by directly invoking `handleUpload` with a `CustomEvent` with the file ref as the `detail`, but I think we should make it a bit cleaner before using in stream.new